### PR TITLE
Add support for Argon2 S2K

### PR DESCRIPTION
**Pull request by lubux**
Adds support for encrypting and decrypting messages or keys with Argon2 S2K as specified in [draft-ietf-openpgp-crypto-refresh](https://datatracker.ietf.org/doc/html/draft-ietf-openpgp-crypto-refresh#section-3.7.1.4).

- Added `Argon2` in the s2k package and refactored it.
- Introduces a new field `S2KConfig` in `packet.Config` that allows configuring S2K. All configurations for S2K should now be done via an S2K configuration linked in this field. The config remains backward compatible to setting `S2KCount` directly in `packet.Config`.
- The public `Encrypt` method exposed in `packet.PrivateKey` for encrypting keys currently does not allow configuring S2K and defaults to the Iterative mode. There is now a private function that allows configuration. (Should be adapted in the future)

### CODE REVIEW

File: openpgp/s2k/s2k.go [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 // Package s2k implements the various OpenPGP string-to-key transforms as
 // specified in RFC 4800 section 3.7.1.
+// Update: Added Argon2 section 3.7.4.4 OpenPGP crypto refresh v8
```
Comment (twiss) [inline openpgp/s2k/s2k.go:7]:
```suggestion
// specified in RFC 4800 section 3.7.1, and Argon2 specified in
// draft-ietf-openpgp-crypto-refresh-08 section 3.7.1.4.
```

File: openpgp/s2k/s2k.go [hunk 2]
```diff
@@ -14,56 +15,123 @@ import (
...
-	// It can be 0 (simple), 1(salted), 3(iterated)
+type S2KType uint8
+
+// Defines the default S2KMode constants
+//
+//	0 (simple), 1(salted), 3(iterated), 4(argon2)
+const (
+	SimpleS2K     S2KType = 0
+	SaltedS2K             = 1
+	IterSaltedS2K         = 3
```
Comment (twiss) [inline openpgp/s2k/s2k.go:29]: I'm personally note the biggest fan of abbreviations, I would just call this `IteratedSaltedS2K`

File: openpgp/s2k/s2k.go [hunk 3]
```diff
@@ -14,56 +15,123 @@ import (
...
+	// Only relevant if S2KMode is not set to s2k.Argon2S2K.
 	// Hash is the default hash function to be used. If
 	// nil, SHA256 is used.
 	Hash crypto.Hash
-	// S2KCount is only used for symmetric encryption. It
+	// Argon2 parameters for S2K (String to Key).
+	// Only relevant if S2KMode is set to s2k.Argon2S2K.
+	// If nil, default parameters are used.
+	// For more details on the choice of parameters, see https://tools.ietf.org/html/rfc9106#section-4.
+	ArgonConf *ArgonConfig
```
Comment (twiss) [inline openpgp/s2k/s2k.go:56]: And similarly, I'd call this `Argon2Config` (both the field and the type)

File: openpgp/s2k/s2k.go [hunk 4]
```diff
@@ -14,56 +15,123 @@ import (
...
+
+func (c *S2KConfig) hash() crypto.Hash {
 	if c == nil || uint(c.Hash) == 0 {
 		return crypto.SHA256
 	}
 
 	return c.Hash
 }
 
+func (c *S2KConfig) ArgonConfig() *ArgonConfig {
```
Comment (twiss) [inline openpgp/s2k/s2k.go:119]: This one you could then call `Argon2()`, for symmetry with `config.AEAD()` and `S2K()`

File: openpgp/s2k/s2k.go [hunk 5]
```diff
@@ -14,56 +15,123 @@ import (
...
+	// For more details on the choice of parameters, see https://tools.ietf.org/html/rfc9106#section-4.
+	ArgonConf *ArgonConfig
+	// Only relevant if S2KMode is set to s2k.IterSaltedS2K.
+	// Iteration count for Iterated S2K (String to Key). It
 	// determines the strength of the passphrase stretching when
 	// the said passphrase is hashed to produce a key. S2KCount
-	// should be between 65536 and 65011712, inclusive. If Config
-	// is nil or S2KCount is 0, the value 16777216 used. Not all
+	// should be between 1024 and 65011712, inclusive. If Config
+	// is nil or S2KCount is 0, the value 65536 used. Not all
```
Comment (twiss) [inline openpgp/s2k/s2k.go:62]: These limits and defaults were set intentionally for security reasons, so they should be kept as is (and `EncodedCount()` below still enforces them)

File: openpgp/s2k/s2k.go [hunk 6]
```diff
@@ -14,56 +15,123 @@ import (
...
+	}
+	return c.ArgonConf
+}
+
+func (c *S2KConfig) Count() int {
+	if c == nil || c.S2KCount == 0 {
+		return 65536
+	}
+	return c.S2KCount
+}
```
Comment (twiss) [inline openpgp/s2k/s2k.go:131]: Do we need this function? It seems unused, and it might be better to have only one way to get the count (via `EncodedCount()` below) - as currently they don't match.

File: openpgp/s2k/s2k.go [hunk 7]
```diff
@@ -14,56 +15,123 @@ import (
...
+//	0 (simple), 1(salted), 3(iterated), 4(argon2)
+const (
+	SimpleS2K     S2KType = 0
+	SaltedS2K             = 1
+	IterSaltedS2K         = 3
+	Argon2S2K             = 4
+	GnuS2K                = 101
+)
+
+const Argon2S2KDefaultNonceSize int = 16
```
Comment (twiss) [inline openpgp/s2k/s2k.go:34]: Since it's not changeable I wouldn't call it "default", and since the spec calls it the salt I'd call this `Argon2SaltSize`

File: openpgp/s2k/s2k.go [hunk 8]
```diff
@@ -169,25 +246,48 @@ func Iterated(out []byte, h hash.Hash, in []byte, salt []byte, count int) {
 	}
 }
 
+func memoryExpToKibibytes(memoryExp uint8) uint32 {
+	return (uint32(1) << memoryExp)
+}
+
+// Argon2Derive writes to out the key derived from the password (in) with the Argon2
+// function (RFC 4880, section 3.7.1.4)
+func Argon2Derive(out []byte, in []byte, salt []byte, passes uint8, paralellism uint8, memoryExp uint8) {
```
Comment (twiss) [inline openpgp/s2k/s2k.go:255]: Since none of the other functions have "derive" in the name I would just call this "Argon2". Also, since it's from the crypto refresh rather than RFC 4880;
```suggestion
// Argon2 writes to out the key derived from the password (in) with the Argon2
// function (crypto refresh, section 3.7.1.4)
func Argon2(out []byte, in []byte, salt []byte, passes uint8, paralellism uint8, memoryExp uint8) {
```

File: openpgp/s2k/s2k.go [hunk 9]
```diff
@@ -14,56 +15,123 @@ import (
...
+	// passes is a parameter in Argon2 to determine the number of iterations
+	// See RFC 4880 Section 3.7.1.4.
+	passes byte
+	// parallelism is a parameter in Argon2 to determine the degree of paralellism
+	// See RFC 4880 Section 3.7.1.4.
+	parallelism byte
+	// memoryExp is a parameter in Argon2 to determine the memory usage
+	// i.e., 2 pow memoryExp kibibytes
+	// See RFC 4880 Section 3.7.1.4.
+	memoryExp byte
```
Comment (twiss) [inline openpgp/s2k/s2k.go:101]:
```suggestion
	// passes is a parameter in Argon2 to determine the number of iterations
	// See the crypto refresh, Section 3.7.1.4.
	passes byte
	// parallelism is a parameter in Argon2 to determine the degree of paralellism
	// See the crypto refresh, Section 3.7.1.4.
	parallelism byte
	// memoryExp is a parameter in Argon2 to determine the memory usage
	// i.e., 2 ** memoryExp kibibytes
	// See the crypto refresh, Section 3.7.1.4.
	memoryExp byte
```

File: openpgp/s2k/s2k.go [hunk 10]
```diff
@@ -80,6 +148,15 @@ func (c *Config) EncodedCount() uint8 {
...
 }
 
+func DefaultArgonConfig() *ArgonConfig {
+	aconf := &ArgonConfig{
+		NumberOfPasses:      3,
+		DegreeOfParallelism: 4,
+		MemoryExponent:      16, // 64 MiB of RAM
+	}
+	return aconf
+}
```
Comment (twiss) [inline openpgp/s2k/s2k.go:158]: Rather than creating a default argon2 config in one go, I would set defaults for the individual parameters, like is done for the other configs (with individual getter functions). That way, people can pass `&ArgonConfig{ MemoryExponent: 32 }` and still get the default value for the other two parameters, for example.

You could move `s2k.Config` and `Argon2Config` to a separate file (e.g. `s2k_config.go`) or even two separate files if you want, to not clutter this one too much with all the getter functions.

File: openpgp/integration_tests/utils_test.go [hunk 11]
```diff
@@ -145,7 +146,7 @@ func randFileHints() *openpgp.FileHints {
 	return &openpgp.FileHints{
 		IsBinary: mathrand.Intn(2) == 0,
 		FileName: string(fileName),
-		ModTime: time.Now(),
+		ModTime:  time.Now(),
```
Comment (twiss) [inline openpgp/integration_tests/utils_test.go:149]: Could you please revert the unrelated whitespace changes in this PR, to make the diff slightly cleaner? Or if you want you can make a separate PR with those first, and then rebase this on top once that's merged.

**Review by twiss — COMMENTED** [review]
Hey :wave: Thanks! See comments below.

File: openpgp/packet/config.go [hunk 12]
```diff
@@ -33,17 +35,10 @@ type Config struct {
...
-	// determines the strength of the passphrase stretching when
-	// the said passphrase is hashed to produce a key. S2KCount
-	// should be between 1024 and 65011712, inclusive. If Config
-	// is nil or S2KCount is 0, the value 65536 used. Not all
-	// values in the above range can be represented. S2KCount will
-	// be rounded up to the next representable value if it cannot
-	// be encoded exactly. When set, it is strongly encrouraged to
-	// use a value that is at least 65536. See RFC 4880 Section
-	// 3.7.1.3.
-	S2KCount int
```
[the remark this answers is not in the document]
Comment (lubux) [reply openpgp/packet/config.go]: Good point, I added it again for backward compatibility.
A dummy `S2KConfig` is created now in `S2K()` if `S2KCount` is set and `S2KConfig` is nil.

**Comment by lubux** [comment]
Thanks for the inputs. I addressed the comments.

File: openpgp/s2k/s2k_config.go [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+
+func (c *ArgonConfig) Passes() uint8 {
+	if c == nil || c.NumberOfPasses == 0 {
+		return 3
+	}
+	return c.NumberOfPasses
+}
+
+func (c *ArgonConfig) Parallelism() uint8 {
+	if c == nil || c.NumberOfPasses == 0 {
```
Comment (twiss) [inline openpgp/s2k/s2k_config.go:111]:
```suggestion
	if c == nil || c.DegreeOfParallelism == 0 {
```

File: openpgp/packet/config.go [hunk 14]
```diff
@@ -33,16 +35,24 @@ type Config struct {
...
 	// CompressionConfig configures the compression settings.
 	CompressionConfig *CompressionConfig
-	// S2KCount is only used for symmetric encryption. It
-	// determines the strength of the passphrase stretching when
+	// S2K (String to Key) config, used for key derivation in the context of secret key encryption
+	// and password-encrypted data.
+	// If nil, the default configuration is used
+	S2KConfig *s2k.Config
+	// Iteration count for Iterated S2K (String to Key). 
+	// Only relevant if sk2.Mode is set to s2k.IteratedSaltedS2K.
```
Comment (twiss) [inline openpgp/packet/config.go:43]: This sort of implies that if you pass `S2KConfig: &s2k.Config{ Mode: IteratedSaltedS2K }` that it's still used, but it's not, so I'd say:
```suggestion
	// Only used if sk2.Mode is nil.
```

**Review by twiss — COMMENTED** [review]
Thanks! Some last comments below. And then, could you update the PR description to match?

File: openpgp/s2k/s2k.go [hunk 15]
```diff
@@ -69,6 +69,31 @@ func decodeCount(c uint8) int {
...
+			return uint8(exp)
+		}
+	}
+
+	return 31
+}
+
+// decodeMemory computes the decoded memory in kibibytes as 2**memoryExponent
+func decodeMemory(memoryExponent uint8) uint32 {
+	return (uint32(1) << memoryExponent)
```
Comment (twiss) [inline openpgp/s2k/s2k.go:94]:
```suggestion
	return uint32(1) << memoryExponent
```

File: openpgp/s2k/s2k_config.go [hunk 16]
```diff
@@ -57,7 +57,9 @@ type Config struct {
 type ArgonConfig struct {
 	NumberOfPasses      uint8
 	DegreeOfParallelism uint8
-	MemoryExponent      uint8
+	// The memory parameter for Argon2 specifies desired memory usage in kibibytes. 
+	// For example memory=64*1024 sets the memory cost to ~64 MB.
+	Memory      		uint32
```
Comment (twiss) [inline openpgp/s2k/s2k_config.go:62]:
```suggestion
	Memory      	    uint32 
```

File: openpgp/s2k/s2k_config.go [hunk 17]
```diff
@@ -108,15 +110,27 @@ func (c *ArgonConfig) Passes() uint8 {
...
 
-func (c *ArgonConfig) Exponent() uint8 {
-	if c == nil || c.MemoryExponent == 0 {
+func (c *ArgonConfig) EncodedMemory() uint8 {
+	if c == nil || c.Memory == 0 {
 		return 16 // 64 MiB of RAM
 	}
-	return c.MemoryExponent
+
+	temp := c.Memory
```
Comment (twiss) [inline openpgp/s2k/s2k_config.go:124]: Very nitpicky but I'd call this `memory`

File: openpgp/s2k/s2k.go (hunk 3 above)
Comment (twiss) [reply openpgp/s2k/s2k.go]: (this one didn't quite get resolved, note the `2` :slightly_smiling_face: since we refer to it by the full name everywhere else I would do so for the config field and type name as well)
